core/blockchain_pruner.go

```diff
@@ -0,0 +1,256 @@
+package core
+
+import (
+	"fmt"
+	"math/big"
+	"math/rand"
+	"os"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/harmony-one/harmony/core/rawdb"
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/utils"
+)
+
+const (
+	// pruneBeaconChainBeforeEpoch Keep 16 epoch
+	pruneBeaconChainBeforeEpoch = 16
+	// pruneBeaconChainBlockBefore Keep 16 epoch blocks
+	pruneBeaconChainBlockBefore = 32768 * pruneBeaconChainBeforeEpoch
+	// maxDeleteBlockOnce max delete block once
+	maxDeleteBlockOnce = 10000
+	// maxMemoryUsage max memory usage is 4 MB
+	maxMemoryUsage = 4 * 1024 * 1024
+	// compactProportion compact proportion is thousandth
+	compactProportion = 1000
+)
+
+type blockchainPruner struct {
+	isPruneBeaconChainRunning int32 // isPruneBeaconChainRunning  must be called atomically
+
+	batchWriter ethdb.Batch
+
+	deleteRange              [][2][]byte
+	startTime                time.Time
+	deletedBlockCount        int
+	deletedValidatorSnapshot int
+	skipValidatorSnapshot    int
+}
+
+func newBlockchainPruner() *blockchainPruner {
+	return &blockchainPruner{}
+}
+
+func (bp *blockchainPruner) initPruneBeaconChain(db ethdb.Database) error {
+	if bp.batchWriter == nil {
```

Comment:
Why don't you just put this code in `newBlockchainPruner`? Lazy initialization is not really needed in this scenario.

Reply:
Removed lazy initialization.